File: src/main/java/org/redrune/utility/game/repository/npc/characteristic/NPCCharacteristic.java
package org.redrune.utility.game.repository.npc.characteristic;

import org.redrune.game.entity.actor.npc.Drop;
import org.redrune.game.entity.actor.npc.data.combat.NPCCombatDefinitions;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Tyluur <itstyluur@icloud.com>
 * @since 8/31/2017
 */
public class NPCCharacteristic {

    /**
     * The map of combat definitions
     */
    private final Map<Integer, NPCCombatDefinitions> combatDefinitions = new HashMap<>();

    /**
     * The map of combat bonuses
     */
    private final Map<Integer, int[]> combatBonuses = new HashMap<>();

    /**
     * The map of drops
     */
    private final Map<Integer, List<Drop>> drops = new HashMap<>();

    /**
     * The map of examines
     */
    private final Map<Integer, String> examines = new HashMap<>();

    /**
     * Adds the examine of an npc
     *
     * @param npcId   The id of the npc
     * @param examine The examine text
     */
    public void addExamine(int npcId, String examine) {
        examines.put(npcId, examine);
    }

    /**
     * Gets an examine
     */
    public String getExamine(int npcId) {
        return examines.get(npcId);
    }

    /**
     * Adds a list of drops
     *
     * @param npcId The id of the npc
     * @param drops The list of drops
     */
    public void addDrops(int npcId, List<Drop> drops) {
        this.drops.put(npcId, drops);
    }

    /**
     * Gets the drops of an npc by its id
     */
    public List<Drop> getDrops(int npcId) {
        return drops.get(npcId);
    }

    /**
     * Adds bonuses to the characteristics
     *
     * @param npcId   The id of the npc
     * @param bonuses The bonuses
     */
    public void addBonuses(int npcId, int[] bonuses) {
        this.combatBonuses.put(npcId, bonuses);
    }

    /**
     * Gets the bonuses of an npc by its id
     */
    public int[] getBonuses(int npcId) {
        return combatBonuses.get(npcId);
    }

    /**
     * Adds combat definitions to the characteristics
     *
     * @param npcId             The id of the npc
     * @param combatDefinitions The combat definitions
     */
    public void addCombatDefinitions(int npcId, NPCCombatDefinitions combatDefinitions) {
        this.combatDefinitions.put(npcId, combatDefinitions);
    }

    /**
     * Gets the combat definitions of an npc by its id
     */
    public NPCCombatDefinitions getCombatDefinitions(int npcId) {
        return combatDefinitions.get(npcId);
    }

    /**
     * Gets the combat definitions of an npc by its id
     */
    public NPCCombatDefinitions getCombatDefinitionsNonNull(int npcId) {
        NPCCombatDefinitions combatDefinitions = this.combatDefinitions.get(npcId);
        if (combatDefinitions == null) {
            combatDefinitions = new NPCCombatDefinitions();
        }
        return combatDefinitions;
    }

    @Override
    public String toString() {
        return "NPCCharacteristic{definitions=" + combatDefinitions.size() + ", combatBonuses=" + combatBonuses.size() + ", drops=" + drops.size() + ", examines=" + examines.size() + "}";
    }
}

File: src/main/kotlin/org/redrune/utility/game/repository/npc/characteristic/NPCCharacteristic.kt
package org.redrune.utility.game.repository.npc.characteristic

import org.redrune.game.entity.actor.npc.Drop
import org.redrune.game.entity.actor.npc.data.combat.NPCCombatDefinitions

/**
 * @author Tyluur <itstyluur@icloud.com>
 * @since 8/31/2017
 */
class NPCCharacteristic {

    /**
     * The map of combat definitions
     */
    private val combatDefinitions: MutableMap<Int, NPCCombatDefinitions> = HashMap()

    /**
     * The map of combat bonuses
     */
    private val combatBonuses: MutableMap<Int, IntArray> = HashMap()

    /**
     * The map of drops
     */
    private val drops: MutableMap<Int, List<Drop>> = HashMap()

    /**
     * The map of examines
     */
    private val examines: MutableMap<Int, String> = HashMap()

    /**
     * Adds the examine of an npc
     *
     * @param npcId   The id of the npc
     * @param examine The examine text
     */
    fun addExamine(npcId: Int, examine: String) {
        examines[npcId] = examine
    }

    /**
     * Gets an examine
     */
    fun getExamine(npcId: Int): String? {
        return examines[npcId]
    }

    /**
     * Adds a list of drops
     *
     * @param npcId The id of the npc
     * @param drops The list of drops
     */
    fun addDrops(npcId: Int, drops: List<Drop>) {
        this.drops[npcId] = drops
    }

    /**
     * Gets the drops of an npc by its id
     */
    fun getDrops(npcId: Int): List<Drop> {
        return drops[npcId]!!
    }

    /**
     * Adds bonuses to the characteristics
     *
     * @param npcId   The id of the npc
     * @param bonuses The bonuses
     */
    fun addBonuses(npcId: Int, bonuses: IntArray) {
        combatBonuses[npcId] = bonuses
    }

    /**
     * Gets the bonuses of an npc by its id
     */
    fun getBonuses(npcId: Int): IntArray? {
        return combatBonuses[npcId]
    }

    /**
     * Adds combat definitions to the characteristics
     *
     * @param npcId             The id of the npc
     * @param combatDefinitions The combat definitions
     */
    fun addCombatDefinitions(npcId: Int, combatDefinitions: NPCCombatDefinitions) {
        this.combatDefinitions[npcId] = combatDefinitions
    }

    /**
     * Gets the combat definitions of an npc by its id
     */
    fun getCombatDefinitions(npcId: Int): NPCCombatDefinitions? {
        return combatDefinitions[npcId]
    }

    /**
     * Gets the combat definitions of an npc by its id
     */
    fun getCombatDefinitionsNonNull(npcId: Int): NPCCombatDefinitions {
        var combatDefinitions = combatDefinitions[npcId]
        if (combatDefinitions == null) {
            combatDefinitions = NPCCombatDefinitions()
        }
        return combatDefinitions
    }

    override fun toString(): String {
        return "NPCCharacteristic{definitions=" + combatDefinitions.size + ", combatBonuses=" + combatBonuses.size + ", drops=" + drops.size + ", examines=" + examines.size + "}"
    }
}